### PR TITLE
add bulkupdateasync

### PR DESCRIPTION
**Pull request by ElizabethOkerio**
### Issues

*This pull request fixes #xxx.*

### Description

*Briefly describe the changes of this pull request.*

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Client/DataServiceContext.cs [hunk 1]
```diff
@@ -2271,6 +2271,74 @@ public virtual DataServiceResponse SaveChanges(SaveChangesOptions options)
...
+        public virtual Task<DataServiceResponse> BulkUpdateAsync<T>(CancellationToken cancellationToken, params T[] objects)
+        {
+            return this.FromAsync((callback, state) => this.BeginBulkUpdate(callback, state, objects), this.EndBulkUpdate, cancellationToken);
+        }
+
+        /// <summary>Asynchronously submits top-level objects to be deep-updated to the data service.</summary>
+        /// <param name="callback">The delegate that is called when a response to the bulk update request is received.</param>
+        /// <param name="state">User-defined state object that is used to pass context data to the callback method.</param>
+        /// <returns>An<see cref="System.IAsyncResult" /> object that is used to track the status of the asynchronous operation. </returns>
+        public virtual IAsyncResult BeginBulkUpdate<T>(AsyncCallback callback, object state, params T[] objects)
```
Comment (habbes) [inline src/Microsoft.OData.Client/DataServiceContext.cs:2318]: Is it worth making these methods `Begin/EndBulkUpdate` methods public. I think existing ones are there for backwards compatibility but is anyone using this pattern in new code when there's `async/await`?

File: src/Microsoft.OData.Client/DataServiceContext.cs [hunk 2]
```diff
@@ -2271,6 +2271,74 @@ public virtual DataServiceResponse SaveChanges(SaveChangesOptions options)
...
+            BulkUpdateSaveResult result = new BulkUpdateSaveResult(this, Util.BulkUpdateMethodName, SaveChangesOptions.BulkUpdate, callback, state);
+            result.BeginBulkUpdateRequest(objects);
+
+            return result;
+        }
+
+        /// <summary>Called to complete the <see cref="BeginBulkUpdate{T}(AsyncCallback, object, T[])"/>.</summary>
+        /// <param name="asyncResult">An <see cref="System.IAsyncResult" /> that represents the status of the asynchronous operation.</param>
+        /// <returns>The DataServiceResponse object that holds the result of the bulk update operation.</returns>
+        public virtual DataServiceResponse EndBulkUpdate(IAsyncResult asyncResult)
```
Comment (habbes) [inline src/Microsoft.OData.Client/DataServiceContext.cs:2334]: Do we need to make this method public?

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs [hunk 3]
```diff
@@ -319,6 +319,57 @@ public void BulkUpdateTwoTopLevelObjects_WithTheSameNestedObject_DeserializesSuc
             Assert.Equal(carTargetEntity, car2TargetEntity);
         }
 
+        [Fact]
```
Comment (habbes) [inline test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs:322]: I think you should have tests that verify the exception handling logic.

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs [hunk 4]
```diff
@@ -99,6 +101,66 @@ protected override Stream ResponseStream
             get { return this.responseStream; }
         }
 
+        /// <summary>
+        /// Asynchronous bulk update request.
+        /// </summary>
+        /// <typeparam name="T"> The type of the top-level objects to be deep-updated.</typeparam>
+        /// <param name="objects"> The top-level objects of the type to be deep updated.</param>
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:108]: fixed.

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs [hunk 5]
```diff
@@ -99,6 +101,66 @@ protected override Stream ResponseStream
...
+                    {
+                        this.HandleCompleted(pereq);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                this.HandleFailure(pereq, e);
+                throw; 
+            }
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:155]: I'm not sure I know why that is but I wanted to make this possible to handle any kind of exception and not just the `InvalidOperationException.`

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs [hunk 6]
```diff
@@ -99,6 +101,66 @@ protected override Stream ResponseStream
...
 
+        /// <summary>
+        /// Asynchronous bulk update request.
+        /// </summary>
+        /// <typeparam name="T"> The type of the top-level objects to be deep-updated.</typeparam>
+        /// <param name="objects"> The top-level objects of the type to be deep updated.</param>
+        internal void BeginBulkUpdateRequest<T>(params T[] objects)
+        {
+            PerRequest pereq = null;
+            ODataRequestMessageWrapper bulkUpdateRequestMessage = null;
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:112]: fixed.

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs [hunk 7]
```diff
@@ -99,6 +101,66 @@ protected override Stream ResponseStream
...
+            }
+
+            BuildDescriptorGraph(this.ChangedEntries, true, objects);
+            
+            try 
+            {
+                bulkUpdateRequestMessage = this.GenerateBulkUpdateRequest();
+                this.Abortable = bulkUpdateRequestMessage;
+
+                if (bulkUpdateRequestMessage != null)
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:126]: Have gotten rid of it.

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs [hunk 8]
```diff
@@ -99,6 +101,66 @@ protected override Stream ResponseStream
...
+                }
+            }
+            catch (Exception e)
+            {
+                this.HandleFailure(pereq, e);
+                throw; 
+            }
+            finally
+            {
+                this.HandleCompleted(pereq);
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:158]: Yes we dispose the request object.

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs [hunk 9]
```diff
@@ -136,6 +198,31 @@ internal void BulkUpdateRequest<T>(params T[] objects)
...
+            base.FinishCurrentChange(pereq);
+
+            Debug.Assert(this.ResponseStream != null, "this.HttpWebResponseStream != null");
+            Debug.Assert((this.ResponseStream as MemoryStream) != null, "(this.HttpWebResponseStream as MemoryStream) != null");
+
+            if (this.ResponseStream.Position != 0)
+            {
+                // Set the stream to the start position and then parse the response and cache it
+                this.ResponseStream.Position = 0;
+                this.HandleBulkUpdateResponse(this.batchResponseMessage, this.ResponseStream);
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:214]: it comes from the base class. It is a property of type `IODataResponseMessage` in the base class that holds the response message.

[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:214]: I didn't want to update the name because it's already used in other places.

File: src/Microsoft.OData.Client/DataServiceContext.cs (hunk 1 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/DataServiceContext.cs:2318]: Ok. I can make them internal.

File: src/Microsoft.OData.Client/DataServiceContext.cs (hunk 2 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/DataServiceContext.cs:2334]: made it internal.

File: src/Microsoft.OData.Client/DataServiceContext.cs [hunk 10]
```diff
@@ -2271,6 +2271,74 @@ public virtual DataServiceResponse SaveChanges(SaveChangesOptions options)
...
+            return this.BulkUpdateAsync(CancellationToken.None, objects);
+        }
+
+        /// <summary>Asynchronously processes a bulk update request</summary>
+        /// <typeparam name="T">The type of top-level objects to be deep updated.</typeparam>
+        /// <param name="objects">The top-level objects of the type to be deep updated.</param>
+        /// <returns>A task representing the <see cref="DataServiceResponse"/> that holds the result of a bulk operation.</returns>
+        public virtual Task<DataServiceResponse> BulkUpdateAsync<T>(CancellationToken cancellationToken, params T[] objects)
+        {
+            return this.FromAsync((callback, state) => this.BeginBulkUpdate(callback, state, objects), this.EndBulkUpdate, cancellationToken);
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/DataServiceContext.cs]: fixed

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs [hunk 11]
```diff
@@ -319,6 +353,108 @@ public void BulkUpdateTwoTopLevelObjects_WithTheSameNestedObject_DeserializesSuc
...
+            var car2TargetEntity = (person2carOperationResponse.Descriptor as LinkDescriptor).Target as Car;
+
+            Assert.Equal(2, response.Count());
+            Assert.Equal("Bing", person1.Name);
+            Assert.Equal("Edge", person2Response.Name);
+            Assert.Equal(carTargetEntity, car2TargetEntity);
+        }
+
+        [Fact]
+        public async Task NullTopLevelObjects_ShouldThrowException()
```
Comment (habbes) [inline test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs:408]: This test name seems inconsistent with what it's actually doing. It doesn't seem to be testing top-level nulls.
Also, in addition to testing for null top-level objects, would also be great to test exceptions that occur inside the async request processing to ensure the async exception handling logic introduced in this PR works as expected.

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (habbes) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:157]: Where is it getting disposed?

I see that we're disposing `perReq` at some point, but looking at the source code of `PerRequest.Dispose()`, it doesn't seem to be disposing it's `PerRequest.Request` object. So it's still not clear to me at what point we're disposing `this.bulkUpdateRequestMessage`.

[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:157]: `perReq` is a wrapper of the request object. When it is disposed it disposes its contents.

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs [hunk 12]
```diff
@@ -99,11 +101,70 @@ protected override Stream ResponseStream
...
+        /// <summary>
+        /// Asynchronous bulk update request.
+        /// </summary>
+        /// <typeparam name="T">The type of the top-level objects to be deep-updated.</typeparam>
+        /// <param name="objects">The top-level objects of the type to be deep updated.</param>
+        internal void BeginBulkUpdateRequest<T>(params T[] objects)
+        {
+            PerRequest pereq = null;
+
+            if (objects == null || objects.Length == 0)
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:113]: why can't check the input at the beginning?

File: src/Microsoft.OData.Client/DataServiceContext.cs [hunk 13]
```diff
@@ -2271,6 +2271,74 @@ public virtual DataServiceResponse SaveChanges(SaveChangesOptions options)
...
+            return result.EndRequest();
+        }
+
+        /// <summary>
+        /// Asynchronously processes a bulk update request. 
+        /// </summary>
+        /// <typeparam name="T">The type of top-level objects to be deep updated.</typeparam>
+        /// <param name="objects">The top-level objects of the type to be deep updated. </param>
+        /// <returns>A task representing the <see cref="DataServiceResponse"/> that holds the result of a bulk operation. </returns>
+        public virtual Task<DataServiceResponse> BulkUpdateAsync<T>(params T[] objects)
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/DataServiceContext.cs:2300]: why don't have async identifier

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs [hunk 14]
```diff
@@ -319,6 +353,108 @@ public void BulkUpdateTwoTopLevelObjects_WithTheSameNestedObject_DeserializesSuc
...
+            this.context.AttachTo("Persons", person);
+            this.context.AttachTo("Persons", person2);
+            this.context.AttachTo("Cars", car);
+
+            this.context.AddLink(person, "Cars", car);
+            this.context.AddLink(person2, "Cars", car);
+
+            DataServiceResponse response = await this.context.BulkUpdateAsync(person, person2);
+
+            var personOperationResponse = response.First() as ChangeOperationResponse;
```
Comment (xuzhg) [inline test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs:392]: can we wrapper those lines into a verifier method? It seems it duplicates.

**Review by xuzhg — DISMISSED** [review]

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs (hunk 11 above)
Comment (KenitoInc) [reply test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs:408]: +1

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs (hunk 12 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:113]: We do check at the beginning. This is an internal method though and can be called from other areas of the project.

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs (hunk 11 above)
Comment (ElizabethOkerio) [reply test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs:408]: Here is an example of a test: https://github.com/OData/odata.net/blob/52487c2c2c620156d53c75909828327133be8210/test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs#L203

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs (hunk 14 above)
Comment (ElizabethOkerio) [reply test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs:416]: We test for different things here. Wrapping these lines might not be of much help.

File: src/Microsoft.OData.Client/DataServiceContext.cs (hunk 13 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/DataServiceContext.cs:2300]: The `FromAsync` method returns a `Task` so we elide the async for efficiency.

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs [hunk 15]
```diff
@@ -193,6 +223,34 @@ public void HandleResponse_Of_AnInvalidResponse_ThrowsException()
...
+            {
+                ID = 1001,
+                Name = "A"
+            };
+
+            this.context.AttachTo("Persons", person);
+
+            this.context.AddRelatedObject(person, "Cars", car);
+
+            await Assert.ThrowsAsync<DataServiceRequestException>(async () => await this.context.BulkUpdateAsync<Person>(person));
```
Comment (habbes) [inline test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs:251]: Is the exception expected to have any details about the error, error message? If so, I think we should also have assertions that verify that.

File: src/Microsoft.OData.Client/BulkUpdateSaveResult.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (habbes) [reply src/Microsoft.OData.Client/BulkUpdateSaveResult.cs:157]: I see. It's strange that it doesn't implement `IDisposable`, seems like design flaw. I wonder whether we may have any resource leaks since at least the underlying `HttpClientRequestMessage` is disposable.

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs (hunk 15 above)
Comment (ElizabethOkerio) [reply test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs]: I've added. The `DateServiceRequestException` has a `DataServiceResponse` property which will be empty whevever an exception is thrown. The exception will have a `Message` and an 'InnerException' but these will differ depending on the deserialization exception thrown.

**Review by habbes — DISMISSED** [review]

**Review by KenitoInc — APPROVED** [review]

**Review by habbes — APPROVED** [review]
